File: crates/cli/src/cli_treemap.rs
use dm_core::tree::Tree;
use dm_core::model::NodeId;

/// ANSI color codes for different depths
const COLORS: &[&str] = &[
    "\x1b[38;5;33m",  // Blue
    "\x1b[38;5;46m",  // Green
    "\x1b[38;5;226m", // Yellow
    "\x1b[38;5;196m", // Red
    "\x1b[38;5;165m", // Magenta
    "\x1b[38;5;51m",  // Cyan
    "\x1b[38;5;208m", // Orange
    "\x1b[38;5;99m",  // Purple
];
const RESET: &str = "\x1b[0m";

/// Cell in the canvas that can hold color info
#[derive(Clone)]
struct Cell {
    ch: char,
    color_index: Option<usize>,
}

impl Cell {
    fn new(ch: char) -> Self {
        Self { ch, color_index: None }
    }

    fn with_color(ch: char, color_index: usize) -> Self {
        Self { ch, color_index: Some(color_index) }
    }
}

/// ASCII treemap visualizer that renders nested rectangles
pub struct TreeMapView {
    width: usize,
    height: usize,
    use_colors: bool,
}

impl TreeMapView {
    pub fn new(width: usize, height: usize) -> Self {
        Self {
            width,
            height,
            use_colors: true,
        }
    }

    pub fn with_colors(mut self, use_colors: bool) -> Self {
        self.use_colors = use_colors;
        self
    }

    fn get_color(&self, depth: usize) -> &'static str {
        if !self.use_colors {
            return "";
        }
        COLORS[depth % COLORS.len()]
    }

    /// Renders a treemap visualization
    /// fs_stats is optional (total_bytes, available_bytes) for showing free space
    /// proportional determines if blocks are sized according to actual disk usage
    pub fn render_tree(&self, tree: &Tree, fs_stats: Option<(u64, u64)>, proportional: bool) -> String {
        let root_id = tree.root();

        // Create a canvas with color info
        let mut canvas = vec![vec![Cell::new(' '); self.width]; self.height];

        // If we have fs_stats, we need to show both used and free space
        if let Some((total_bytes, available_bytes)) = fs_stats {
            let used_bytes = tree.node(root_id).total_bytes();

            // Render as two side-by-side sections
            self.render_with_free_space(
                tree,
                root_id,
                &mut canvas,
                used_bytes,
                available_bytes,
                total_bytes,
                proportional,
            );
        } else {
            // Render the tree recursively (normal mode)
            self.render_node(
                tree,
                root_id,
                &mut canvas,
                0,
                0,
                self.width,
                self.height,
                0,
            );
        }

        // Convert canvas to colored string
        self.canvas_to_string(&canvas)
    }

    fn canvas_to_string(&self, canvas: &[Vec<Cell>]) -> String {
        let mut result = String::new();
        let mut current_color: Option<usize> = None;

        for row in canvas {
            for cell in row {
                if self.use_colors {
                    if cell.color_index != current_color {
                        // Color changed, emit new color code
                        if let Some(color_idx) = cell.color_index {
                            result.push_str(COLORS[color_idx % COLORS.len()]);
                        } else {
                            result.push_str(RESET);
                        }
                        current_color = cell.color_index;
                    }
                }
                result.push(cell.ch);
            }
            if self.use_colors && current_color.is_some() {
                result.push_str(RESET);
                current_color = None;
            }
            result.push('\n');
        }

        result
    }

    fn render_with_free_space(
        &self,
        tree: &Tree,
        root_id: NodeId,
        canvas: &mut Vec<Vec<Cell>>,
        used_bytes: u64,
        available_bytes: u64,
        total_bytes: u64,
        proportional: bool,
    ) {
        let (used_width, free_width) = if proportional {
            // Calculate proportional widths based on used vs available
            let comparison_total = used_bytes + available_bytes;
            let used_ratio = if comparison_total > 0 {
                used_bytes as f64 / comparison_total as f64
            } else {
                0.5
            };

            let used_width = (self.width as f64 * used_ratio).ceil().max(1.0) as usize;
            let free_width = self.width.saturating_sub(used_width);

            (used_width, free_width)
        } else {
            // Non-proportional: split width evenly
            let half_width = self.width / 2;
            (half_width, self.width.saturating_sub(half_width))
        };

        // Render used space (the tree structure)
        if used_width > 3 {
            self.render_node(
                tree,
                root_id,
                canvas,
                0,
                0,
                used_width,
                self.height,
                0,
            );
        }

        // Render free space as a simple box
        if free_width > 3 {
            let free_x = used_width;
            self.draw_free_space_box(canvas, free_x, 0, free_width, self.height, available_bytes);
        }
    }

    fn draw_free_space_box(
        &self,
        canvas: &mut Vec<Vec<Cell>>,
        x: usize,
        y: usize,
        width: usize,
        height: usize,
        free_bytes: u64,
    ) {
        if width < 3 || height < 3 {
            return;
        }

        // Use a gray color for free space (color index 7 = purple, let's use that)
        let color_idx = 7; // Or we could add a new gray color

        // Draw border
        self.draw_box(canvas, x, y, width, height, color_idx);

        // Add label
        if width > 10 && height > 2 {
            let label = format!("Free ({})", Self::format_size(free_bytes));
            let label_width = width.saturating_sub(4);
            self.draw_text(canvas, x + 2, y + 1, &label, label_width, color_idx);
        }
    }

    fn render_node(
        &self,
        tree: &Tree,
        node_id: NodeId,
        canvas: &mut Vec<Vec<Cell>>,
        x: usize,
        y: usize,
        width: usize,
        height: usize,
        depth: usize,
    ) {
        if width < 3 || height < 3 {
            return;
        }

        let node = tree.node(node_id);
        let children = node.children();

        // Draw border with color
        self.draw_box(canvas, x, y, width, height, depth);

        // Add label (only if there's enough space)
        if width > 4 && height > 2 {
            let name = node.name();
            let size = node.total_bytes();
            let label_width = width.saturating_sub(4);
            let label = format!("{} ({})", Self::truncate_name(name, label_width), Self::format_size(size));
            self.draw_text(canvas, x + 2, y + 1, &label, label_width, depth);
        }

        // Calculate layout for children
        if !children.is_empty() && height > 5 && width > 6 {
            let inner_x = x.saturating_add(2).min(canvas[0].len().saturating_sub(1));
            let inner_y = y.saturating_add(3).min(canvas.len().saturating_sub(1));
            let inner_width = width.saturating_sub(4);
            let inner_height = height.saturating_sub(5);

            if inner_width > 0 && inner_height > 0 {
                self.layout_children(
                    tree,
                    children,
                    canvas,
                    inner_x,
                    inner_y,
                    inner_width,
                    inner_height,
                    depth + 1,
                );
            }
        }
    }

    fn layout_children(
        &self,
        tree: &Tree,
        children: &[NodeId],
        canvas: &mut Vec<Vec<Cell>>,
        x: usize,
        y: usize,
        width: usize,
        height: usize,
        depth: usize,
    ) {
        if children.is_empty() || width == 0 || height == 0 {
            return;
        }

        // Calculate total size
        let total_size: u64 = children
            .iter()
            .map(|&id| tree.node(id).total_bytes())
            .sum();

        if total_size == 0 {
            return;
        }

        // Use squarified treemap algorithm (simplified)
        let mut remaining_children: Vec<(NodeId, u64)> = children
            .iter()
            .map(|&id| (id, tree.node(id).total_bytes()))
            .filter(|(_, size)| *size > 0)
            .collect();

        remaining_children.sort_by(|a, b| b.1.cmp(&a.1)); // Sort by size descending

        let mut current_x = x;
        let mut current_y = y;
        let mut remaining_width = width;
        let mut remaining_height = height;

        for (child_id, child_size) in remaining_children {
            if remaining_width == 0 || remaining_height == 0 {
                break;
            }

            let ratio = child_size as f64 / total_size as f64;

            // Decide direction based on aspect ratio
            let (child_width, child_height) = if remaining_width > remaining_height {
                // Split horizontally
                let w = (remaining_width as f64 * ratio).max(1.0).min(remaining_width as f64) as usize;
                let w = w.clamp(1, remaining_width);
                (w, remaining_height)
            } else {
                // Split vertically
                let h = (remaining_height as f64 * ratio).max(1.0).min(remaining_height as f64) as usize;
                let h = h.clamp(1, remaining_height);
                (remaining_width, h)
            };

            self.render_node(
                tree,
                child_id,
                canvas,
                current_x,
                current_y,
                child_width,
                child_height,
                depth,
            );

            // Update position for next child
            if remaining_width > remaining_height {
                current_x += child_width;
                remaining_width = remaining_width.saturating_sub(child_width);
            } else {
                current_y += child_height;
                remaining_height = remaining_height.saturating_sub(child_height);
            }
        }
    }

    fn draw_box(
        &self,
        canvas: &mut Vec<Vec<Cell>>,
        x: usize,
        y: usize,
        width: usize,
        height: usize,
        depth: usize,
    ) {
        if width < 2 || height < 2 {
            return;
        }

        // Use different border styles for different depths
        let (tl, tr, bl, br, h, v) = match depth % 4 {
            0 => ('┌', '┐', '└', '┘', '─', '│'),
            1 => ('╔', '╗', '╚', '╝', '═', '║'),
            2 => ('╭', '╮', '╰', '╯', '─', '│'),
            _ => ('┏', '┓', '┗', '┛', '━', '┃'),
        };

        let max_x = (x + width - 1).min(canvas[0].len() - 1);
        let max_y = (y + height - 1).min(canvas.len() - 1);

        let color_index = if self.use_colors { Some(depth) } else { None };

        // Top and bottom
        for i in (x + 1)..max_x {
            if y < canvas.len() && i < canvas[0].len() {
                canvas[y][i] = if let Some(ci) = color_index {
                    Cell::with_color(h, ci)
                } else {
                    Cell::new(h)
                };
            }
            if max_y < canvas.len() && i < canvas[0].len() {
                canvas[max_y][i] = if let Some(ci) = color_index {
                    Cell::with_color(h, ci)
                } else {
                    Cell::new(h)
                };
            }
        }

        // Left and right
        for i in (y + 1)..max_y {
            if i < canvas.len() && x < canvas[0].len() {
                canvas[i][x] = if let Some(ci) = color_index {
                    Cell::with_color(v, ci)
                } else {
                    Cell::new(v)
                };
            }
            if i < canvas.len() && max_x < canvas[0].len() {
                canvas[i][max_x] = if let Some(ci) = color_index {
                    Cell::with_color(v, ci)
                } else {
                    Cell::new(v)
                };
            }
        }

        // Corners
        if y < canvas.len() && x < canvas[0].len() {
            canvas[y][x] = if let Some(ci) = color_index {
                Cell::with_color(tl, ci)
            } else {
                Cell::new(tl)
            };
        }
        if y < canvas.len() && max_x < canvas[0].len() {
            canvas[y][max_x] = if let Some(ci) = color_index {
                Cell::with_color(tr, ci)
            } else {
                Cell::new(tr)
            };
        }
        if max_y < canvas.len() && x < canvas[0].len() {
            canvas[max_y][x] = if let Some(ci) = color_index {
                Cell::with_color(bl, ci)
            } else {
                Cell::new(bl)
            };
        }
        if max_y < canvas.len() && max_x < canvas[0].len() {
            canvas[max_y][max_x] = if let Some(ci) = color_index {
                Cell::with_color(br, ci)
            } else {
                Cell::new(br)
            };
        }
    }

    fn draw_text(
        &self,
        canvas: &mut Vec<Vec<Cell>>,
        x: usize,
        y: usize,
        text: &str,
        max_width: usize,
        depth: usize,
    ) {
        if y >= canvas.len() || x >= canvas[0].len() || max_width == 0 {
            return;
        }

        let available_width = canvas[0].len().saturating_sub(x);
        let write_width = max_width.min(available_width);

        for (i, ch) in text.chars().take(write_width).enumerate() {
            let pos = x + i;
            if pos < canvas[0].len() {
                canvas[y][pos] = if self.use_colors {
                    Cell::with_color(ch, depth)
                } else {
                    Cell::new(ch)
                };
            }
        }
    }

    fn truncate_name(name: &str, max_len: usize) -> String {
        if name.len() <= max_len {
            name.to_string()
        } else if max_len < 3 {
            name.chars().take(max_len).collect()
        } else {
            format!("{}...", name.chars().take(max_len - 3).collect::<String>())
        }
    }

    fn format_size(bytes: u64) -> String {
        const KB: u64 = 1024;
        const MB: u64 = KB * 1024;
        const GB: u64 = MB * 1024;
        const TB: u64 = GB * 1024;

        if bytes >= TB {
            format!("{:.1}T", bytes as f64 / TB as f64)
        } else if bytes >= GB {
            format!("{:.1}G", bytes as f64 / GB as f64)
        } else if bytes >= MB {
            format!("{:.1}M", bytes as f64 / MB as f64)
        } else if bytes >= KB {
            format!("{:.1}K", bytes as f64 / KB as f64)
        } else {
            format!("{}B", bytes)
        }
    }
}

impl Default for TreeMapView {
    fn default() -> Self {
        Self::new(120, 40)
    }
}

File: nodejs/cli/cli_treemap.js
const { FileMetaData } = require('../dm_core/top_files');

const COLORS = [
  '\u001b[38;5;33m',
  '\u001b[38;5;46m',
  '\u001b[38;5;226m',
  '\u001b[38;5;196m',
  '\u001b[38;5;165m',
  '\u001b[38;5;51m',
  '\u001b[38;5;208m',
  '\u001b[38;5;99m',
];
const RESET = '\u001b[0m';

class Cell {
  constructor(ch, colorIndex = null) {
    this.ch = ch;
    this.colorIndex = colorIndex;
  }

  static new(ch) {
    return new Cell(ch, null);
  }

  static withColor(ch, colorIndex) {
    return new Cell(ch, colorIndex);
  }
}

class TreeMapView {
  constructor(width, height) {
    this.width = width;
    this.height = height;
    this.useColors = true;
  }

  withColors(useColors) {
    this.useColors = useColors;
    return this;
  }

  getColor(depth) {
    if (!this.useColors) {
      return '';
    }
    return COLORS[depth % COLORS.length];
  }

  renderTree(tree, fsStats, proportional) {
    const rootId = tree.rootId();
    const canvas = Array.from({ length: this.height }, () =>
      Array.from({ length: this.width }, () => Cell.new(' '))
    );

    if (fsStats) {
      const [totalBytes, availableBytes] = fsStats;
      const usedBytes = tree.node(rootId).totalBytes();
      this.renderWithFreeSpace(
        tree,
        rootId,
        canvas,
        usedBytes,
        availableBytes,
        totalBytes,
        proportional
      );
    } else {
      this.renderNode(tree, rootId, canvas, 0, 0, this.width, this.height, 0);
    }

    return this.canvasToString(canvas);
  }

  canvasToString(canvas) {
    let result = '';
    let currentColor = null;

    for (const row of canvas) {
      for (const cell of row) {
        if (this.useColors) {
          if (cell.colorIndex !== currentColor) {
            if (cell.colorIndex !== null && cell.colorIndex !== undefined) {
              result += COLORS[cell.colorIndex % COLORS.length];
            } else {
              result += RESET;
            }
            currentColor = cell.colorIndex;
          }
        }
        result += cell.ch;
      }
      if (this.useColors && currentColor !== null) {
        result += RESET;
        currentColor = null;
      }
      result += '\n';
    }

    return result;
  }

  renderWithFreeSpace(
    tree,
    rootId,
    canvas,
    usedBytes,
    availableBytes,
    totalBytes,
    proportional
  ) {
    let usedWidth;
    let freeWidth;

    if (proportional) {
      const comparisonTotal = usedBytes + availableBytes;
      const usedRatio = comparisonTotal > 0 ? usedBytes / comparisonTotal : 0.5;
      usedWidth = Math.round(this.width * usedRatio);
      freeWidth = Math.max(0, this.width - usedWidth);

      const MIN_WIDTH = 15;
      if (usedBytes > 0 && usedWidth < MIN_WIDTH) {
        usedWidth = Math.min(MIN_WIDTH, this.width - MIN_WIDTH);
        freeWidth = Math.max(0, this.width - usedWidth);
      }
      if (availableBytes > 0 && freeWidth < MIN_WIDTH) {
        freeWidth = Math.min(MIN_WIDTH, this.width - MIN_WIDTH);
        usedWidth = Math.max(0, this.width - freeWidth);
      }
    } else {
      const halfWidth = Math.floor(this.width / 2);
      usedWidth = halfWidth;
      freeWidth = Math.max(0, this.width - halfWidth);
    }

    if (usedWidth >= 3) {
      this.renderNode(tree, rootId, canvas, 0, 0, usedWidth, this.height, 0);
    }

    if (freeWidth >= 3) {
      const freeX = usedWidth;
      this.drawFreeSpaceBox(canvas, freeX, 0, freeWidth, this.height, availableBytes);
    }
  }

  drawFreeSpaceBox(canvas, x, y, width, height, freeBytes) {
    if (width < 3 || height < 3) {
      return;
    }

    const colorIdx = 7;
    this.drawBox(canvas, x, y, width, height, colorIdx);

    if (width > 10 && height > 2) {
      const label = `Free (${TreeMapView.formatSize(freeBytes)})`;
      const labelWidth = Math.max(0, width - 4);
      this.drawText(canvas, x + 2, y + 1, label, labelWidth, colorIdx);
    }
  }

  renderNode(tree, nodeId, canvas, x, y, width, height, depth) {
    if (width < 3 || height < 3) {
      return;
    }

    const node = tree.node(nodeId);
    const children = node.childrenValue();

    this.drawBox(canvas, x, y, width, height, depth);

    let nextContentY = y + 1;
    if (width > 4 && height > 2) {
      const name = node.nameValue();
      const size = node.totalBytes();
      const labelWidth = Math.max(0, width - 4);
      const sizeLabel = TreeMapView.formatSize(size);
      const sizeWithParens = ` (${sizeLabel})`;
      const availableForName = Math.max(0, labelWidth - sizeWithParens.length);
      const nameLabel = TreeMapView.truncateName(name, availableForName);
      const label = `${nameLabel}${sizeWithParens}`;
      this.drawText(canvas, x + 2, nextContentY, label, labelWidth, depth);
      nextContentY += 1;
    }

    const items = [];
    for (const childId of children) {
      items.push({ kind: 'folder', id: childId });
    }
    const topFiles = node.topFiles().toSortedVecDesc();
    for (const file of topFiles) {
      items.push({ kind: 'file', file });
    }

    const innerHeight = height - (nextContentY - y + 2);
    if (items.length > 0 && height > nextContentY - y + 2 && width > 4) {
      const innerX = x + 2;
      const innerY = nextContentY;
      const innerWidth = Math.max(0, width - 4);

      if (innerWidth > 0 && innerHeight > 0) {
        this.layoutChildren(
          tree,
          items,
          canvas,
          innerX,
          innerY,
          innerWidth,
          innerHeight,
          depth + 1
        );
      }
    }
  }

  renderFile(canvas, x, y, width, height, depth, name, size) {
    if (width < 1 || height < 1) {
      return;
    }

    if (width >= 5 && height >= 3) {
      this.drawBox(canvas, x, y, width, height, depth);
      const labelWidth = Math.max(0, width - 4);
      const sizeLabel = TreeMapView.formatSize(size);
      const sizeWithParens = ` (${sizeLabel})`;
      const availableForName = Math.max(0, labelWidth - sizeWithParens.length);
      const nameLabel = TreeMapView.truncateName(name, availableForName);
      const label = `${nameLabel}${sizeWithParens}`;
      this.drawText(canvas, x + 2, y + 1, label, labelWidth, depth);
    } else {
      const fillCh = (() => {
        switch (depth % 4) {
          case 0:
            return '\u2591';
          case 1:
            return '\u2592';
          case 2:
            return '\u2593';
          default:
            return '\u2588';
        }
      })();

      for (let row = y; row < y + height; row += 1) {
        for (let col = x; col < x + width; col += 1) {
          if (row < canvas.length && col < canvas[0].length) {
            if (canvas[row][col].ch === ' ') {
              canvas[row][col] = this.useColors
                ? Cell.withColor(fillCh, depth)
                : Cell.new(fillCh);
            }
          }
        }
      }

      if (width >= 5 && height >= 1) {
        const label = TreeMapView.truncateName(name, Math.max(0, width - 2));
        const labelWithDash = `- ${label}`;
        this.drawText(canvas, x, y, labelWithDash, width, depth);
      } else if (width >= 2 && height >= 1) {
        const label = TreeMapView.truncateName(name, width);
        this.drawText(canvas, x, y, label, width, depth);
      }
    }
  }

  layoutChildren(tree, items, canvas, x, y, width, height, depth) {
    if (items.length === 0 || width === 0 || height === 0) {
      return [];
    }

    const sizedItems = items
      .map((item) => ({
        item,
        size: item.kind === 'folder' ? tree.node(item.id).totalBytes() : item.file.size,
      }))
      .filter((entry) => entry.size > 0)
      .sort((a, b) => b.size - a.size);

    const totalDisplaySize = sizedItems.reduce((sum, entry) => sum + entry.size, 0);
    if (totalDisplaySize === 0) {
      return [];
    }

    let currentX = x;
    let currentY = y;
    let remainingWidth = width;
    let remainingHeight = height;
    let remainingTotalSize = totalDisplaySize;
    const skipped = [];

    for (let i = 0; i < sizedItems.length; i += 1) {
      const { item, size } = sizedItems[i];

      if (remainingWidth === 0 || remainingHeight === 0 || remainingTotalSize === 0) {
        skipped.push(...sizedItems.slice(i).map((entry) => entry.item));
        break;
      }

      const ratio = size / remainingTotalSize;
      let itemWidth;
      let itemHeight;

      if (remainingWidth > remainingHeight * 1.5) {
        itemWidth = Math.round(remainingWidth * ratio);
        itemWidth = Math.max(1, Math.min(itemWidth, remainingWidth));
        itemHeight = remainingHeight;
      } else {
        itemHeight = Math.round(remainingHeight * ratio);
        itemHeight = Math.max(1, Math.min(itemHeight, remainingHeight));
        itemWidth = remainingWidth;
      }

      if (itemWidth < 1 || itemHeight < 1) {
        skipped.push(item);
      } else if (item.kind === 'folder') {
        this.renderNode(tree, item.id, canvas, currentX, currentY, itemWidth, itemHeight, depth);
      } else if (item.kind === 'file') {
        this.renderFile(
          canvas,
          currentX,
          currentY,
          itemWidth,
          itemHeight,
          depth,
          item.file.name,
          item.file.size
        );
      }

      if (remainingWidth > remainingHeight * 1.5) {
        currentX += itemWidth;
        remainingWidth = Math.max(0, remainingWidth - itemWidth);
      } else {
        currentY += itemHeight;
        remainingHeight = Math.max(0, remainingHeight - itemHeight);
      }
      remainingTotalSize = Math.max(0, remainingTotalSize - size);
    }

    return skipped;
  }

  drawBox(canvas, x, y, width, height, depth) {
    if (width < 2 || height < 2) {
      return;
    }

    const styles = [
      ['\u250c', '\u2510', '\u2514', '\u2518', '\u2500', '\u2502'],
      ['\u2554', '\u2557', '\u255a', '\u255d', '\u2550', '\u2551'],
      ['\u256d', '\u256e', '\u2570', '\u256f', '\u2500', '\u2502'],
      ['\u250f', '\u2513', '\u2517', '\u251b', '\u2501', '\u2503'],
    ];

    const [tl, tr, bl, br, h, v] = styles[depth % 4];

    const maxX = Math.min(x + width - 1, canvas[0].length - 1);
    const maxY = Math.min(y + height - 1, canvas.length - 1);

    const colorIndex = this.useColors ? depth : null;

    for (let i = x + 1; i < maxX; i += 1) {
      if (y < canvas.length && i < canvas[0].length) {
        canvas[y][i] = colorIndex !== null ? Cell.withColor(h, colorIndex) : Cell.new(h);
      }
      if (maxY < canvas.length && i < canvas[0].length) {
        canvas[maxY][i] = colorIndex !== null ? Cell.withColor(h, colorIndex) : Cell.new(h);
      }
    }

    for (let i = y + 1; i < maxY; i += 1) {
      if (i < canvas.length && x < canvas[0].length) {
        canvas[i][x] = colorIndex !== null ? Cell.withColor(v, colorIndex) : Cell.new(v);
      }
      if (i < canvas.length && maxX < canvas[0].length) {
        canvas[i][maxX] = colorIndex !== null ? Cell.withColor(v, colorIndex) : Cell.new(v);
      }
    }

    if (y < canvas.length && x < canvas[0].length) {
      canvas[y][x] = colorIndex !== null ? Cell.withColor(tl, colorIndex) : Cell.new(tl);
    }
    if (y < canvas.length && maxX < canvas[0].length) {
      canvas[y][maxX] = colorIndex !== null ? Cell.withColor(tr, colorIndex) : Cell.new(tr);
    }
    if (maxY < canvas.length && x < canvas[0].length) {
      canvas[maxY][x] = colorIndex !== null ? Cell.withColor(bl, colorIndex) : Cell.new(bl);
    }
    if (maxY < canvas.length && maxX < canvas[0].length) {
      canvas[maxY][maxX] = colorIndex !== null ? Cell.withColor(br, colorIndex) : Cell.new(br);
    }
  }

  drawText(canvas, x, y, text, maxWidth, depth) {
    if (y >= canvas.length || x >= canvas[0].length || maxWidth === 0) {
      return;
    }

    const availableWidth = Math.max(0, canvas[0].length - x);
    const writeWidth = Math.min(maxWidth, availableWidth);

    let idx = 0;
    for (const ch of text) {
      if (idx >= writeWidth) {
        break;
      }
      const pos = x + idx;
      if (pos < canvas[0].length) {
        canvas[y][pos] = this.useColors ? Cell.withColor(ch, depth) : Cell.new(ch);
      }
      idx += 1;
    }
  }

  static truncateName(name, maxLen) {
    if (name.length <= maxLen) {
      return name;
    }
    if (maxLen < 1) {
      return '';
    }
    if (maxLen < 3) {
      return name.slice(0, maxLen);
    }
    const take = Math.max(0, maxLen - 2);
    return `${name.slice(0, take)}..`;
  }

  static formatSize(bytes) {
    const KB = 1024;
    const MB = KB * 1024;
    const GB = MB * 1024;
    const TB = GB * 1024;

    if (bytes >= TB) {
      return `${(bytes / TB).toFixed(1)}T`;
    }
    if (bytes >= GB) {
      return `${(bytes / GB).toFixed(1)}G`;
    }
    if (bytes >= MB) {
      return `${(bytes / MB).toFixed(1)}M`;
    }
    if (bytes >= KB) {
      return `${(bytes / KB).toFixed(1)}K`;
    }
    return `${bytes}B`;
  }
}

module.exports = {
  TreeMapView,
};
